### PR TITLE
Support for optgroups

### PR DESCRIPTION
**Pull request by bgrundmann**
What it says on the tin.  I chose not to support django forms API (where the choices are interpreted differently if the values are lists or tuples), even so it leads to pretty callsites. 

 In practice I found the proposed API more eh practical ;-)

Existing test cases pass and the adjusted form kitchen sink does what you expect.

BUT note the comment on the existing choice_tuples method.

### CODE REVIEW

File: iommi/form.py [hunk 1]
```diff
@@ -838,6 +844,25 @@ def choice_tuples(self):
...
+        if self.choice_to_optgroup is None:
+            return [(None, self.choice_tuples)]
+        else:
+            groups = []
+            current_group_name = None
+            current_group = []
+            groups.append((current_group_name, current_group))
+            for choice_tuple in self.choice_tuples:
+                choice = choice_tuple[0]
+                group_name = self.choice_to_optgroup(choice)
```
Comment (boxed) [inline iommi/form.py:858]: We should send in `**self.iommi_evaluate_parameters()` here too, and add `**_` to your example above.

Comment (bgrundmann) [reply iommi/form.py:858]: Sure I'll get right on it... a while later ... Wait what?  evaluate, evaluate_strict, signature_from_kwargs... What is all this magic?

Comment (boxed) [reply iommi/form.py:858]: I have begun thinking about how to write a document to explain all that. It might take me a little while..

Comment (bgrundmann) [reply iommi/form.py:858]: Great.  Looking forward.

File: iommi/form.py [hunk 2]
```diff
@@ -827,6 +831,8 @@ def choice_to_options_selected(self):
 
     @property
     def choice_tuples(self):
+        # Why does this cache exist?  choice_tuples is only called once to render
```
Comment (bgrundmann) [inline iommi/form.py:834]: Does anybody have an answer to this?

Comment (boxed) [reply iommi/form.py:834]: I think you're right. It was probably useful at some point in the past and wasn't removed when it wasn't needed anymore...

File: iommi/form.py (hunk 1 above)
Comment (jlubcke) [reply iommi/form.py:858]: The `.iommi_evaluate_parameters()` is a function to get a set of "reasonable" parameters to all those callbacks. When we traverse the whole tree of components all things visited on the way down has the option to add themselves (and/or other practical parameters) to a collection of parameters that is sent to most (should be all) callbacks. That is why you see signatures like `my_callback(field, **_)`. The **_ is those you don't need and a safeguard against more things beeing included in the future. 

Things already in the parameters include 'request' (set from the root of the traversal), 'traversable' that is the closest node in the graph, 'page', 'table', 'column' etc.

But all the gathering is done by the `Traversable.bind()` machinery and the interface is to override the `
own_evaluate_parameters()` method to add something to the set of parameters, and to remember to add the result of `.iommi_evaluate_parameters()` when you invoke callbacks. (typically with the help of `evaluate` or `evaluate_strict`)

Comment (bgrundmann) [reply iommi/form.py]: Thanks for that explanation.  I think this patch now does what I want.  Let me know if you want any further changes.

**Pull request merged**